### PR TITLE
KP-734 [Zip] Consolidate into single method for Compress Archive

### PR DESCRIPTION
**Pull request by xkcm**
[**KP-734 in Jira**](https://kaholo.atlassian.net/browse/KP-734)

### CODE REVIEW

**Review by kenskyfish — DISMISSED** [review]
I tested thoroughly, looks great. Thanks.

File: config.json [hunk 1]
```diff
@@ -5,17 +5,17 @@
   "execProgram": "node",
   "main": "app.js",
   "imgUrl": "logo.png",
-  "version": "1.1.2",
+  "version": "2.0.5",
```
Comment (graddaniel) [inline config.json]: It's the first occurence of v2, thus it should be `2.0.0`

File: app.js [hunk 2]
```diff
@@ -1,66 +1,61 @@
...
+  const absoluteTargetPaths = targetPaths.map((targetPath) => path.resolve(targetPath));
+  const absoluteIgnoredPaths = (ignoredPaths || []).map((ignorePath) => path.resolve(ignorePath));
+  await validatePaths(...absoluteTargetPaths, path.dirname(absoluteArchivePath));
+
+  const archivePathExists = await pathExists(absoluteArchivePath);
+  if (overwrite && archivePathExists) {
+    await fs.rm(absoluteArchivePath, { recursive: true });
+  } else if (!overwrite && archivePathExists) {
+    throw new Error(`File ${archivePath} already exists. Please select the "Overwrite Existing Zip Archive" parameter to overwrite.`);
   }
```
Comment (graddaniel) [inline app.js]:
```suggestion
  if (!overwrite && archivePathExists) {
    throw new Error(`File ${archivePath} already exists. Please select the "Overwrite Existing Zip Archive" parameter to overwrite.`);
  }
  
  if (archivePathExists) {
    await fs.rm(absoluteArchivePath, { recursive: true });
  }
```

This way the same algorithmic complexity is maintained (1 condition evaluation at best, 2 at worst case), but we separate the functionalities - unhappy path from optional action.

File: app.js [hunk 3]
```diff
@@ -1,66 +1,61 @@
...
+  const absoluteArchivePath = path.resolve(archivePath);
+  const absoluteDestinationPath = path.resolve(destinationPath);
+  await validatePaths(absoluteArchivePath);
+
+  const destinationPathExists = await pathExists(absoluteDestinationPath);
+  if (clearExtractionPath && destinationPathExists) {
+    await fs.rm(absoluteDestinationPath, { recursive: true });
+  } else if (!clearExtractionPath && destinationPathExists) {
+    throw new Error(`Path ${destinationPath} exists. Please select the "Clear Extraction Path" parameter to overwrite.`);
   }
```
Comment (graddaniel) [inline app.js]:
```suggestion
  if (!clearExtractionPath && destinationPathExists) {
    throw new Error(`Path ${destinationPath} exists. Please select the "Clear Extraction Path" parameter to overwrite.`);
  }
  
  if (destinationPathExists) {
    await fs.rm(absoluteDestinationPath, { recursive: true });
  }
```
same here

File: helpers.js [hunk 4]
```diff
@@ -1,4 +1,16 @@
...
+const { async: AsyncStreamZip } = require("node-stream-zip");
+const {
+  basename: resolveBaseName,
+  relative: resolveRelativePath,
+  resolve: resolvePath,
+} = require("path");
+const Archiver = require("archiver");
+const fs = require("fs");
+
+const { lstat, access, mkdir } = fs.promises;
```
Comment (graddaniel) [inline helpers.js]: Is this approach documented anywhere, or did it just happen to work?

File: helpers.js [hunk 5]
```diff
@@ -9,6 +21,84 @@ async function pathExists(path) {
...
+    }
+    const pathStat = await lstat(resolvePath(rootPath, path));
+    return pathStat.isDirectory() ? [`${path}/${MATCH_ALL_PATH}`, path] : path;
+  });
+
+  const resolvedPaths = await Promise.all(resolvedPathPromises);
+  return resolvedPaths.flat();
+}
+
 module.exports = {
```
Comment (graddaniel) [inline helpers.js]: This whole file isn't generic, it revolves around paths and zip files. There's definitely a better name to describe it altogether (or separately), than `helpers.js`. Let's not use generics, when it's possible to use specifics.

**Review by graddaniel — CHANGES_REQUESTED** [review]

File: helpers.js (hunk 4 above)
Comment (xkcm) [reply helpers.js]: https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fs_promises_api

**Review by graddaniel — APPROVED** [review]

**Pull request merged**